Fixes part of #1096 : CrashLogger [Blocked on #1241]

Sarthak2601 opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->
## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->
Fixes part of #1096 
- Added Crashlytics Wrapper for custom logging of exceptions 
- Implemented custom logging in the Utility module's DataProviders
- Tested implementation in DataProvidersTest
- Added google-services.json files of a test project to the app and utility modules
- Added Firebase related dependencies. 

## Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

Sarthak2601 committed:
crashlytics wrapper

Sarthak2601 committed:
wrapper check

vinitamurthi committed:
Test implementation of crashlytics mocks

Sarthak2601 committed:
Crashlytics Wrapper in utility

Sarthak2601 committed:
nits

vinitamurthi commented on app/google-services.json:
```diff
@@ -0,0 +1,69 @@
+  "project_info": {
+    "project_number": "409851798978",
+    "firebase_url": "https://oppia-android-40f74.firebaseio.com",
+    "project_id": "oppia-android-40f74",
```
Just to confirm, this is for testing purposes only right? Maybe its worth adding a comment to mention this is our test project (@BenHenning FYI when we actually release, we would want to change these IDs to the project you are creating)

vinitamurthi replied:
Also, @Sarthak2601 can you comment on all parts of this json file that would need an update when we have to release (I know we would have to change project_id, and maybe storage_bucket..what about the client params below?)

Sarthak2601 replied:
I think we should replace the whole json file at the time of release because almost everything will change.

BenHenning replied:
We probably shouldn't check this in until it's associated with our actual project.

Further, is there anything we need to be worried about regarding exposing the information here to the world? Ie, does any of this need to be private? If so, we probably need to create two different projects: one that developers use, and one that's used by the shipped app.

vinitamurthi replied:
The problem is that the app won't build if this file doesn't exist. These params are dummy for the test world, what we could do is just keep an empty or minimal json file just so that the app builds

vinitamurthi replied:
This has been moved to another PR, resolving this

vinitamurthi commented on app/google-services.json:
```diff
@@ -0,0 +1,69 @@
+      ],
+      "api_key": [
+        {
+          "current_key": "AIzaSyDGAkJdWwOt7eLwh47YB4D-mwcZTB4nPi4"
```
These are test keys as far as I understand so it should be safe to include them in our github repo correct? @BenHenning FYI again here we would probably have to change them to the actual release keys before releasing
@Sarthak2601 this is another param we should comment above mentioning that it needs a change on release

Sarthak2601 replied:
@vinitamurthi can't comment as json doesn't allow that.

vinitamurthi commented on app/google-services.json:
```diff
@@ -0,0 +1,69 @@
+    }
+  ],
+  "configuration_version": "1"
+}
```
Add a new line at the EOF

Sarthak2601 replied:
Done

vinitamurthi commented on utility/google-services.json:
```diff
@@ -0,0 +1,69 @@
+{
```
Same comments that were in the app module json apply here. Also add a comment above explaining why we need two json files.

Sarthak2601 replied:
File deleted because of no usage.

vinitamurthi commented on utility/google-services.json:
```diff
@@ -0,0 +1,69 @@
+    }
+  ],
+  "configuration_version": "1"
+}
```
Add a new line at the EOF

Sarthak2601 replied:
File deleted because of no usage.

vinitamurthi commented on utility/google-services.json:
```diff
@@ -0,0 +1,69 @@
+{
+  "project_info": {
```
Can you format this file using https://jsonformatter.org/ ?

Sarthak2601 replied:
File deleted because of no usage.

vinitamurthi commented on app/google-services.json:
```diff
@@ -0,0 +1,69 @@
+{
```
Can you format this file using https://jsonformatter.org/ ?

Sarthak2601 replied:
Format returned same version.

vinitamurthi commented on utility/src/main/java/org/oppia/util/crashlytics/CrashlyticsModule.kt:
```diff
@@ -0,0 +1,15 @@
+import javax.inject.Singleton
+
+@Module
+class CrashlyticsModule {
```
As we discussed earlier, rename this to FirebaseModule not crashlytics. Also change the directory name to firebase

Sarthak2601 replied:
Done

vinitamurthi commented on utility/src/test/java/org/oppia/util/data/DataProvidersTest.kt:
```diff
@@ -57,6 +61,12 @@ private const val COMBINED_STR_VALUE_02 = "I used to be indecisive. At least I t
+   * Returns Mockito.any() as nullable type to avoid java.lang.IllegalStateException when
+   * null is returned.
+   */
+  fun <T> any(): T = Mockito.any<T>()
```
Can you create an issue mentioning that we need to move these functions into a helper class? You can assign that to me, and add a comment with the issue here in the following format:
`TODO(<issue number>): Add a MockitoHelper class to handle nullable versions of all mockito matchers`

Sarthak2601 replied:
Done 
Issue created - #1233

rt4914 replied:
Few nits:
```
 // TODO(#1233): Add a MockitoHelper class to handle nullable versions of all mockito matchers.
  /**
   * Returns Mockito.any() as nullable type to avoid java.lang.IllegalStateException when
   * null is returned.
   */
  fun <T> any(): T = Mockito.any<T>()
```
1. The TODO should comment on top instead of bottom.
2. There should no space between `TODO` and `(`
3. Full stop (.) at the end of sentence.

Sarthak2601 replied:
Done

vinitamurthi replied:
I think you don't need this any() function anymore

Sarthak2601 replied:
Done

vinitamurthi commented on utility/src/test/java/org/oppia/util/data/DataProvidersTest.kt:
```diff
@@ -92,6 +102,7 @@ class DataProvidersTest {
   @Before
   fun setUp() {
     setUpTestApplicationComponent()
+    MockitoAnnotations.initMocks(this)
```
I don't think we need this anymore since we are manually mocking the class below (i.e. doing `Mockito.mock(CrashlyticsWrapper::class` below)

Sarthak2601 replied:
Done

vinitamurthi reviewed (changes requested):
Overall looks good, a couple of nits primarily! @BenHenning @rt4914 can you'll also take a look?

Sarthak2601 committed:
nits

Sarthak2601 committed:
nit

Sarthak2601 committed:
Firebase config file and dependency changes

Sarthak2601 committed:
nits

BenHenning commented on utility/src/test/java/org/oppia/util/data/DataProvidersTest.kt:
```diff
@@ -407,13 +417,13 @@ class DataProvidersTest {
     verify(mockStringLiveDataObserver).onChanged(stringResultCaptor.capture())
     assertThat(stringResultCaptor.value.isFailure()).isTrue()
     assertThat(stringResultCaptor.value.getErrorOrNull()).isInstanceOf(IllegalStateException::class.java)
+    verify(TestFirebaseModule.mockCrashlyticsWrapper, atLeastOnce()).logException(any())
```
This is a new behavior being tested (that we log an exception for failing data providers. Can you add these as distinct tests? Otherwise, it slightly muddles the behaviors being tested here.

Also, I don't think we need to test in every single place. Just having a single failure test for each interesting DataProvider case is probably sufficient.

Sarthak2601 replied:
Actually they are added only in those test cases which were going into the try-catch block and actually hitting our function i.e the three dataProvider cases. The reason behind this is now the test case can test the error handling in totality. 

If we create separate test cases just for those three data providers and remove this from here, we'll not be able to verify if our failure test cases actually do all the functionalities that its required to do.

BenHenning replied:
I see you added one new test per my feedback, but what I meant was one test for each of the data provider methods being affected (rather than changing all of the failing tests).

Sarthak2601 replied:
Done, PTAL.

BenHenning commented on utility/src/main/java/org/oppia/util/firebase/CrashlyticsWrapper.kt:
```diff
@@ -0,0 +1,27 @@
+
+/** Controller for providing custom crash reporting to Firebase Crashlytics */
+@Singleton
+open class CrashlyticsWrapper @Inject constructor(
```
We should generally discourage inheritance. Could we make this an interface, instead, and mock that? Then, we could provide an implementation of the wrapper that actually associates with the FirebaseCrashlytics implementation, and another that provides a fake implementation. You could use Mockito for the latter, but it might be cleaner to build an actual fake from which you can retrieve the logged exceptions & messages.

Sarthak2601 replied:
Done, PTAL

BenHenning commented on app/google-services.json:
```diff
@@ -0,0 +1,40 @@
+{
```
To keep this PR simpler, can we move the JSON file to its own, dedicated PR that this one is blocked on? I feel like this PR is doing a couple different things. Keeping it separate will make the review a bit easier, and let us focus on the individual aspects of each PR.

Sarthak2601 replied:
Okay so in terms of the security thing, the risk is that if we put up the config file our api key will be exposed. However, that API key can only be used to fiddle with our backend (if any) on our firebase console. 
Since we don't have any, it won't be an issue.  

The reason I am keen on merging it together is, without the config file the app won't build thereby creating a halt in everyone's work. Also it's a config of a test project, we'll have to change it whenever we decide to release. 

Other alternative - We can tell our developers to add their own config files as a part of setting up the project. The downside to this is, all of them will have to additionally go to Firebase, create a project, register the app and put the config file to be able to run it. It might be okay for the existing devs, but I don't think that will be the case for the new contributors.

vinitamurthi replied:
This has moved to another PR so I am resolving this

BenHenning commented on utility/src/main/java/org/oppia/util/data/DataProviders.kt:
```diff
@@ -20,7 +21,8 @@ import org.oppia.util.threading.BackgroundDispatcher
   @BackgroundDispatcher private val backgroundDispatcher: CoroutineDispatcher,
-  private val asyncDataSubscriptionManager: AsyncDataSubscriptionManager
+  private val asyncDataSubscriptionManager: AsyncDataSubscriptionManager,
+  private val crashlyticsWrapper: CrashlyticsWrapper
```
A note on naming: while it a wrapper, if you change this to be an interface (per a later comment of mine) it may be clearer to name it based on its anticipated functionality, rather than its implementation. It seems to be a CrashLogger, and right now we're using Crashlytics as the implementation for that logger. I suggest naming it based on its usage rather than its implementation.

Sarthak2601 replied:
Renamed to crashLogger

BenHenning replied:
Sorry, I left a comment recently suggesting we use ExceptionLogger instead. I should have thought about this a bit more. It occurred to me now that this isn't actually meant for logging crashes, but instead cases when the app doesn't crash. Even if we add fatal exception logging later, 'ExceptionLogger' still seems a bit more generic.

BenHenning commented on utility/src/main/java/org/oppia/util/firebase/CrashlyticsWrapper.kt:
```diff
@@ -0,0 +1,27 @@
+import javax.inject.Inject
+import javax.inject.Singleton
+
+/** Controller for providing custom crash reporting to Firebase Crashlytics */
```
Regarding the documentation, this isn't really a controller. Maybe 'logger' would be clearer (per my comment above re: 'CrashLogger').

Sarthak2601 replied:
Done

BenHenning commented on utility/src/main/java/org/oppia/util/firebase/FirebaseModule.kt:
```diff
@@ -0,0 +1,15 @@
+class FirebaseModule {
+  @Singleton
+  @Provides
+  fun provideFirebaseCrashlytics(): FirebaseCrashlytics {
```
Maybe instead provide an implementation of the wrapper/logger class and pass in FirebaseCrashlytics (I'm not sure if we actually want the crashlytics object in the DI graph--we probably want to hide it within one of our implementations to keep future migrations/refactors a bit simpler).

Sarthak2601 replied:
Done

BenHenning commented on utility/src/main/java/org/oppia/util/firebase/CrashlyticsWrapper.kt:
```diff
@@ -0,0 +1,27 @@
+  }
+
+  /** Logs a custom log message which can be put alongside a crash report to Firebase Crashlytics */
+  open fun logMessage(message: String) {
```
When will this be used? It doesn't seem to be in this PR.

Sarthak2601 replied:
Made for future usage (if any). We can probably use it to provide a better insight in our crash reports.

BenHenning commented on utility/src/main/java/org/oppia/util/firebase/CrashlyticsWrapper.kt:
```diff
@@ -0,0 +1,27 @@
+  }
+
+  /** Sets up a user identifier which is attached in every crash report to Firebase Crashlytics */
+  open fun setUserIdentifier(identifier: String) {
```
One of the intents of this project was to not associate logs with an individual user--how is this meant to be used?

Sarthak2601 replied:
Okay, so both the logMessage and setUserIdentifier are kept for future usages. 
One use case of setUserIdentifier can be providing the countries of our users. We can then possibly see if our app is crashing in a particular region.

BenHenning replied:
In general, we should only implement what we specifically need for a particular piece of the project or in anticipation of work soon-to-be done. We can always add new functionality later once it's determined to be needed.

BenHenning reviewed (changes requested):
Thanks @Sarthak2601! I left a bunch of comments. Please send the PR back my way once you have a chance to address them. Also, feel free to follow up with me regarding questions if you have any.

rt4914 commented on app/build.gradle:
```diff
@@ -85,6 +86,7 @@ dependencies {
       'com.google.dagger:dagger:2.24',
       'com.google.firebase:firebase-analytics-ktx:17.4.2',
       'com.google.firebase:firebase-crashlytics:17.0.0',
+      'com.google.firebase:firebase-core:17.4.2',
```
Nit: Rearrange them in alphabetical order:
```
      'com.google.firebase:firebase-core:17.4.2',
      'com.google.firebase:firebase-crashlytics:17.0.0',
```

Sarthak2601 replied:
Done

rt4914 commented on utility/build.gradle:
```diff
@@ -50,6 +50,9 @@ dependencies {
       'com.google.dagger:dagger:2.24',
+      'com.google.firebase:firebase-analytics-ktx:17.4.2',
+      'com.google.firebase:firebase-crashlytics:17.0.0',
+      'com.google.firebase:firebase-core:17.4.2',
```
Nit: Rearrange them in alphabetical order:
```
      'com.google.firebase:firebase-core:17.4.2',
      'com.google.firebase:firebase-crashlytics:17.0.0',
```

Sarthak2601 replied:
Done

rt4914 commented on utility/src/main/java/org/oppia/util/firebase/CrashlyticsWrapper.kt:
```diff
@@ -0,0 +1,27 @@
+  private var firebaseCrashlytics: FirebaseCrashlytics
+) {
+
+  /** Logs a custom non-fatal exception to Firebase Crashlytics */
```
In general each comment statement should have `.` at the end. Ditto elsewhere in this file.

Sarthak2601 replied:
Done both the things

rt4914 reviewed (changes requested):
@Sarthak2601 PTAL. Thanks.

Sarthak2601 committed:
nits | renaming of wrapper

Sarthak2601 committed:
nits + interface

Sarthak2601 committed:
ktlint nits

Sarthak2601 committed:
json deletion for seperate PR

vinitamurthi commented on utility/src/main/java/org/oppia/util/firebase/CrashLogger.kt:
```diff
@@ -0,0 +1,27 @@
+  override fun logException(exception: Exception) {
+    firebaseCrashlytics.recordException(exception)
+  }
+
```
Let's remove this for now

Sarthak2601 replied:
Done

Sarthak2601 committed:
new testing approach

Sarthak2601 committed:
Merge branch 'develop' into crashlytics_wrapper

Sarthak2601 committed:
new testing approach

Sarthak2601 commented:
@vinitamurthi @BenHenning @rt4914 
- Added a FakeCrashLogger and TestFirebaseModule for faking crash logging.
- Wrapped up FirebaseCrashlytics inside provideCrashLogger() in FirebaseModule. 
- Introduced a new test in the DataProvidersTest which tests our fakes. 

PTAL, thanks :)

vinitamurthi commented on testing/src/main/java/org/oppia/testing/FakeCrashLogger.kt:
```diff
@@ -0,0 +1,21 @@
+
+  fun getMostRecentException(): Exception {
+    val size = exceptionList.size
+    return exceptionList[size - 1]
```
You probably want to check if size is > 0 here before returning

Sarthak2601 replied:
Done

vinitamurthi replied:
This would return -1 if the size = 0, you should check if size > 0

vinitamurthi commented on testing/src/main/java/org/oppia/testing/FakeCrashLogger.kt:
```diff
@@ -0,0 +1,21 @@
+    return exceptionList[size - 1]
+  }
+
+  fun cleanExceptionList() {
```
Call this clearAllExceptions instead

Sarthak2601 replied:
Done

vinitamurthi commented on utility/src/main/java/org/oppia/util/firebase/FirebaseModule.kt:
```diff
@@ -0,0 +1,15 @@
+import javax.inject.Singleton
+
+@Module
+class FirebaseModule {
```
Let's not call this FirebaseModule, maybe LogReportingModule?  (I am generally bad with names so @rt4914 @BenHenning do you have any suggestions )

rt4914 replied:
Even I am not good with naming but `LogReportingModule` sounds better than `FirebaseModule`.

Also @Sarthak2601 Please add KDoc for every new class or public functions.

Sarthak2601 replied:
Done

vinitamurthi commented on testing/src/main/java/org/oppia/testing/TestFirebaseModule.kt:
```diff
@@ -0,0 +1,20 @@
+import org.oppia.util.firebase.CrashLogger
+
+@Module
+class TestFirebaseModule {
```
TestLogReportingModule maybe? (Depends on the name you use for the module below)

Sarthak2601 replied:
Done

vinitamurthi reviewed (changes requested):
Just a few naming things, overall looks good!

rt4914 reviewed (approved):
LGTM, resolved my previous comments.

Sarthak2601 committed:
nits

vinitamurthi commented on utility/src/main/java/org/oppia/util/firebase/CrashLogger.kt:
```diff
@@ -0,0 +1,11 @@
+ */
+interface CrashLogger {
+
+  fun logException(exception: Exception)
```
Need a java doc comment for this function

Sarthak2601 replied:
Done.

vinitamurthi commented on utility/src/test/java/org/oppia/util/data/DataProvidersTest.kt:
```diff
@@ -2842,6 +2845,25 @@ class DataProvidersTest {
+    dataProviders.convertToLiveData(dataProvider).observeForever(mockIntLiveDataObserver)
+    testCoroutineDispatchers.advanceUntilIdle()
+
+    assertThat(fakeCrashLogger.getMostRecentException()).isInstanceOf(IllegalStateException::class.java)
```
Assert the message of the exception as well to ensure that we are logging the IllegalStateException properly. Looks like the message is "Base failure" so you can add a statement verifying that `fakeCrashLogger.getMostRecentException().message` is equal to "Base Failure"

Sarthak2601 replied:
Done

vinitamurthi reviewed (approved):
LGTM! Just two comments about adding another assertion in a test and adding a comment to the interface.

vinitamurthi commented on testing/src/main/java/org/oppia/testing/FakeCrashLogger.kt:
```diff
@@ -0,0 +1,30 @@
+    return if (size > 0) {
+      exceptionList[size - 1]
+    } else {
+      exceptionList[size]
```
This will return an ArrayIndexOutOfBoundsException , your size is 0 so it will try to return the 0th element which doesn't exist. You should throw an error here not send exceptionList[size]

Sarthak2601 replied:
Yes, my bad :( 
Modified it to throw `NullPointerException("No element present")`, will it be good enough ?

vinitamurthi replied:
This looks fine

Sarthak2601 committed:
changes

vinitamurthi commented:
LGTM! Once the json file goes in you can merge this after @BenHenning's approval!

BenHenning commented on testing/src/main/java/org/oppia/testing/FakeCrashLogger.kt:
```diff
@@ -0,0 +1,30 @@
+
+  /** This is used to get the most recent exception present in the list of exceptions. */
+  fun getMostRecentException(): Exception {
+    val size = exceptionList.size
```
A more Kotlin idiomatic way to do this would be:

```kotlin
fun getMostRecentException(): Exception = exceptionList.last()
```

See documentation: https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/last.html.

Sarthak2601 replied:
Done.

BenHenning commented on testing/src/main/java/org/oppia/testing/FakeCrashLogger.kt:
```diff
@@ -0,0 +1,30 @@
+    }
+  }
+
+  /** This is used to clear all the exceptions present in the list. */
```
Method KDocs should be written as a truncated verb phrase to complete the statement: "This method ___", e.g.: ``Clears all the exceptions currently logged.``

Ditto elsewhere.

Note that I also removed 'the list' since that's an implementation detail. Documentation comments should focus on external users of the object. They want to know what functionality is available & what it does, not how it's implemented since that doesn't help solve whatever problem they're trying to use the component to solve.

Sarthak2601 replied:
Done, PTAL

BenHenning commented on utility/src/main/java/org/oppia/util/firebase/CrashLoggerImplementation.kt:
```diff
@@ -0,0 +1,16 @@
+
+/** Logger for providing custom crash reporting to Firebase Crashlytics. */
+@Singleton
+open class CrashLoggerImplementation(
```
This is one implementation of crash logger, the fake is another. In this case, this isn't a fake logger and instead relies on a real backing implementation (Firebase). I suggest naming this accordingly, e.g. ``FirebaseCrashLogger.``

Sarthak2601 replied:
Done.

BenHenning commented on testing/src/main/java/org/oppia/testing/FakeCrashLogger.kt:
```diff
@@ -0,0 +1,30 @@
+package org.oppia.testing
+
+import java.lang.Exception
+import org.oppia.testing.TestLogReportingModule.Companion.exceptionList
```
We should never use static mutable state. If you want to share state across multiple components that are injecting this logger, you can define the exception list as a field of the class and mark the class as ``@Singleton`` so that its instance is retained across injections. You'll need to make this an injectable class for that to work, e.g.:

```kotlin
@Singleton
class FakeCrashLogger @Inject constructor() : CrashLogger {
```

Sarthak2601 replied:
Done.

BenHenning commented on utility/src/main/java/org/oppia/util/firebase/CrashLogger.kt:
```diff
@@ -0,0 +1,14 @@
+import java.lang.Exception
+
+/**
+ * Interface that enables exception logging.
```
The 'Interface that' is unnecessary, and slightly distracting (calling out details in a documentation comment emphasizes their importance--is it actually important that this is an interface?). Also, this doesn't describe what the logger is meant to do. Consider an alernative:

```Logger for tracking caught exceptions. Note that this utility may later upload the exceptions to remote services, or log the exception to a file on disk.```

Sarthak2601 replied:
Done, PTAL

BenHenning commented on utility/src/main/java/org/oppia/util/firebase/CrashLogger.kt:
```diff
@@ -0,0 +1,14 @@
+/**
+ * Interface that enables exception logging.
+ */
+interface CrashLogger {
```
Is this really a crash logger, or just an exception logger? A crash logger suggest it's specific to fatal crashes.

Sarthak2601 replied:
Renamed to ExceptionLogger

BenHenning commented on utility/src/main/java/org/oppia/util/firebase/CrashLogger.kt:
```diff
@@ -0,0 +1,14 @@
+ */
+interface CrashLogger {
+
+  /** @param [exception] : the non fatal exception that is mostly caught in try-catch blocks.
```
This KDoc is formatted incorrectly (see https://developer.android.com/kotlin/style-guide#formatting_2), and it's missing a summary.

Sarthak2601 replied:
Done, PTAL.

BenHenning commented on utility/src/main/java/org/oppia/util/firebase/CrashLogger.kt:
```diff
@@ -0,0 +1,14 @@
+interface CrashLogger {
+
+  /** @param [exception] : the non fatal exception that is mostly caught in try-catch blocks.
+   * This function basically facilitates custom exception logging to Crashlytics.
```
Similar to other comments--Crashlytics isn't relevant to this component from a user's perspective. Users just want to log exceptions. You can note like I suggested in an earlier comment that exceptions may be logged to remote services or locally on disk, but _how_ it does that should be left up to the implementation (consider that the fake can't actually satisfy this requirement since it doesn't use Crashlytics).

Sarthak2601 replied:
Done.

BenHenning commented on utility/src/main/java/org/oppia/util/firebase/CrashLoggerImplementation.kt:
```diff
@@ -0,0 +1,16 @@
+import javax.inject.Singleton
+
+/** Logger for providing custom crash reporting to Firebase Crashlytics. */
+@Singleton
```
Since this class isn't injected, this scope won't do anything. It can be removed.

Sarthak2601 replied:
Done

BenHenning commented on utility/src/main/java/org/oppia/util/firebase/CrashLoggerImplementation.kt:
```diff
@@ -0,0 +1,16 @@
+
+/** Logger for providing custom crash reporting to Firebase Crashlytics. */
+@Singleton
+open class CrashLoggerImplementation(
```
This class isn't actually being extended, so suggest removing the ``open`` property to prefer the default state of immutability.

Sarthak2601 replied:
Done.

BenHenning commented on utility/src/test/java/org/oppia/util/data/DataProvidersTest.kt:
```diff
@@ -2865,6 +2845,27 @@ class DataProvidersTest {
+    testCoroutineDispatchers.advanceUntilIdle()
+
+    assertThat(fakeCrashLogger.getMostRecentException()).isInstanceOf(IllegalStateException::class.java)
+    assertThat(fakeCrashLogger.getMostRecentException().message).matches("Base failure")
```
The slightly more Truth idiomatic way to do this would be:

```kotlin
assertThat(fakeCrashLogger.getMostRecentException()).hasMessageThat().contains("Base failure")
```

Also, it's completely fine to extract the exception to a local variable to shorten these assertion lines.

Sarthak2601 replied:
Done.

BenHenning commented on utility/src/main/java/org/oppia/util/firebase/LogReportingModule.kt:
```diff
@@ -5,11 +5,14 @@ import dagger.Module
 import javax.inject.Singleton
 
+/**
+ * Provides crash logging dependencies.
```
Perhaps: ``Provides Firebase-specific logging implementations.``

That sets up the module for later Firebase logging work. Also, this KDoc can fit on one line, so prefer that, e.g.:

```kotlin
/** Doc comment is short and on one line. */
```

Sarthak2601 replied:
Done.

BenHenning commented on testing/src/main/java/org/oppia/testing/TestLogReportingModule.kt:
```diff
@@ -0,0 +1,23 @@
+  }
+
+  @Provides
+  @Singleton
```
If you make the fake crash logger an injectable ``@Singleton``, you don't need to redo that here since calling the provider at each injection site for ``CrashLogger`` will reuse the same fake crash logger. Instead, you can bind it to keep it simple:

```kotlin
@Module
interface TestLogReportingModule {
  @Binds
  fun bindFakeCrashLogger(FakeCrashLogger impl): CrashLogger
}
```

Sarthak2601 replied:
Done.

BenHenning commented on testing/src/main/java/org/oppia/testing/FakeCrashLogger.kt:
```diff
@@ -0,0 +1,30 @@
+/**
+ * A test specific fake crash logger that adds the logged exception to a list of exceptions.
+ */
+class FakeCrashLogger : CrashLogger {
```
Although it's trivial, I suggest adding a very basic test suite to verify the contract of the fake crash logger itself.

Sarthak2601 replied:
Done.

BenHenning reviewed (changes requested):
Thanks @Sarthak2601! I think the fake approach looks quite good. I had a bunch of thoughts on structure & documentation since this is likely a pattern we're going to use a few times throughout this project, and possible as an example for future projects (I think this is the first fake we're introducing in the codebase).

BenHenning commented on utility/src/main/java/org/oppia/util/firebase/CrashLogger.kt:
```diff
@@ -0,0 +1,14 @@
+package org.oppia.util.firebase
```
Suggest moving this to a non-Firebase specific project since the interface isn't part of the implementation. Maybe org.oppia.util.logging?

Sarthak2601 replied:
Done.

BenHenning commented on utility/src/main/java/org/oppia/util/firebase/CrashLoggerImplementation.kt:
```diff
@@ -0,0 +1,16 @@
+package org.oppia.util.firebase
```
Similarly, a Firebase-specific implementation might go under ``org.oppia.util.logging.firebase``.

Sarthak2601 replied:
Done.

Sarthak2601 committed:
changes

Sarthak2601 committed:
testing related changes.

Sarthak2601 committed:
Merge branch 'develop' into crashlytics_wrapper

Sarthak2601 committed:
Tests.

Sarthak2601 committed:
Merge remote-tracking branch 'origin/crashlytics_wrapper' into crashlytics_wrapper

Sarthak2601 committed:
module addition in domain files.

Sarthak2601 committed:
Merge branch 'develop' into crashlytics_wrapper

BenHenning commented on testing/src/main/java/org/oppia/testing/TestLogReportingModule.kt:
```diff
@@ -1,23 +1,15 @@
-    return FakeCrashLogger()
-  }
+  @Binds
+  abstract fun bindFakeCrashLogger(fakeExceptionLogger: FakeExceptionLogger): ExceptionLogger
```
Optional, but I think you could make the class an interface instead since it only contains abstract methods.

Sarthak2601 replied:
Done.

BenHenning commented on testing/src/main/java/org/oppia/testing/TestLogReportingModule.kt:
```diff
@@ -1,23 +1,15 @@
-    return FakeCrashLogger()
-  }
+  @Binds
+  abstract fun bindFakeCrashLogger(fakeExceptionLogger: FakeExceptionLogger): ExceptionLogger
```
Nit: ``bindFakeExceptionLogger``

Sarthak2601 replied:
Done.

BenHenning commented on testing/src/main/java/org/oppia/testing/TestLogReportingModule.kt:
```diff
@@ -1,23 +1,15 @@
+import org.oppia.util.logging.ExceptionLogger
 
 /**
  * Provides fake crash logging dependencies.
```
Please update the KDoc to match the class's purpose. In this case, it seems to be for fake log reporting dependencies.

Sarthak2601 replied:
Done.

BenHenning commented on testing/src/test/java/org/oppia/testing/FakeExceptionLoggerTest.kt:
```diff
@@ -0,0 +1,194 @@
+    }
+
+    assertThat(exception).isInstanceOf(NoSuchElementException::class.java)
+    assertThat(exception).hasMessageThat().contains("List is empty.")
```
This is technically an implementation detail of Kotlin we can probably omit. It's a good idea to check the message of the failure, but in this case the context and exception is pretty specific, so maybe just keep the check for ``NoSuchElementException`` type.

Sarthak2601 replied:
Done.

BenHenning commented on testing/src/main/java/org/oppia/testing/FakeExceptionLogger.kt:
```diff
@@ -0,0 +1,22 @@
+/** A test specific fake for the exception logger. */
+@Singleton
+class FakeExceptionLogger @Inject constructor() : ExceptionLogger {
+  var exceptionList = ArrayList<Exception>()
```
A more Kotlin-idiomatic structure to use is a mutable list, e.g.:

```kotlin
private val exceptionList = mutableListOf<Exception>()
```

Also, prefer using ``val`` for read-only fields, and prefer to minimize access to members.

Sarthak2601 replied:
Done.

BenHenning commented on testing/src/test/java/org/oppia/testing/FakeExceptionLoggerTest.kt:
```diff
@@ -0,0 +1,194 @@
+  @Test
+  fun testFakeExceptionLogger_clearAllExceptions_returnsEmptyList() {
+    fakeExceptionLogger.clearAllExceptions()
+    val exceptionList = fakeExceptionLogger.exceptionList
```
We shouldn't access the implementations of the class (prefer to test the API of a class, not its implementation details). If you think it's useful for tests to be able to query whether an exception has been logged, you could introduce a method like ``hasExceptionsLogged()`` and test that. Or, you could infer that the exception list is inferred through other methods (like how ``getMostRecentException`` throws if called after ``clearAllExceptions`` is called).

The purpose of tests should be to verify that a method of the public API behaves correctly under certain conditions, not that it has a particular implementation. We should be able to switch the full implementation of a class and not change any tests if they are correctly only testing the API of the class.

Ditto elsewhere here.

Sarthak2601 replied:
Okay, done :)

BenHenning commented on testing/src/test/java/org/oppia/testing/FakeExceptionLoggerTest.kt:
```diff
@@ -0,0 +1,194 @@
+
+    fun inject(fakeExceptionLoggerTest: FakeExceptionLoggerTest)
+  }
+}
```
Nit: please add EOF newline.

Sarthak2601 replied:
Done.

BenHenning commented on utility/src/test/java/org/oppia/util/data/DataProvidersTest.kt:
```diff
@@ -2860,10 +2860,49 @@ class DataProvidersTest {
+  @InternalCoroutinesApi
+  @ExperimentalCoroutinesApi
+  fun testCombine_combinerThrowsException_deliversFailure_logsException() {
+    fakeExceptionLogger.clearAllExceptions()
```
Here & elsewhere: do we actually need to clear the exceptions? There should be a new fake logger test/test since you're scoping the state of the fake to singleton, and we create a new application component for each test.

Sarthak2601 replied:
Yes I realised we don't need it, so modified accordingly.

BenHenning reviewed (approved):
Thanks @Sarthak2601! I mostly just have nits. I think @vinitamurthi can verify that my comments in this pass are addressed if she has an opportunity to before I return next week.

Sarthak2601 committed:
Merge branch 'develop' into crashlytics_wrapper

Sarthak2601 committed:
lint fix

Sarthak2601 committed:
changes.

Sarthak2601 committed:
testLogReportingModule in app module

Sarthak2601 committed:
Robolectric tests passing.

Sarthak2601 committed:
Merge branch 'develop' into crashlytics_wrapper

Sarthak2601 committed:
espresso tests running as well.

Sarthak2601 committed:
Merge remote-tracking branch 'origin/crashlytics_wrapper' into crashlytics_wrapper

Sarthak2601 committed:
non-flaky correction

vinitamurthi commented:
LGTM, I think all of ben's comments were addressed and checks have passed so merging this PR.

Sarthak2601 commented:
Thanks @vinitamurthi :)
